### PR TITLE
[OptApp] Restructure 12 - Add Solid Isotropic Material with Penalization method Control

### PR DESCRIPTION
**Pull request by sunethwarna**
**📝 Description**
This PR adds `SIMPControl` - Solid Isotropic Material with Penalization method as a control mechanism so that `OptimizationApplication` can do topology optimization.

Todo:
- [ ] Add test

**🆕 Changelog**
- Added `SimpControl`
- Added `SimpUtils`

### CODE REVIEW

File: applications/OptimizationApplication/custom_utilities/controls/simp_utils.cpp [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+// Application includes
+
+// Include base h
+#include "simp_utils.h"
+
+namespace Kratos
+{
+
+double SimpUtils::ProjectForward(
```
Comment (FMeister) [inline applications/OptimizationApplication/custom_utilities/controls/simp_utils.cpp:28]: I feel like the names could be more descriptive. What is X and onto what is it projected?

File: applications/OptimizationApplication/custom_utilities/controls/simp_utils.cpp [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    const IndexType x2 = x1 + 1;
+
+    const double y1 = rYLimits[x1];
+    const double y2 = rYLimits[x2];
+    const double pow_val = -2.0 * Beta * (X - (x1 + x2) / 2.0);
+
+    return (y2 - y1) / (std::pow(1 + std::exp(pow_val), PenaltyFactor)) + y1;
+}
+
+double SimpUtils::ProjectBackward(
```
Comment (FMeister) [inline applications/OptimizationApplication/custom_utilities/controls/simp_utils.cpp:44]: Same here ...

File: applications/OptimizationApplication/custom_utilities/controls/simp_utils.cpp [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+// Project includes
+#include "includes/model_part.h"
+#include "utilities/variable_utils.h"
+#include "utilities/parallel_utilities.h"
+
+// Application includes
+
+// Include base h
+#include "simp_utils.h"
```
Comment (FMeister) [inline applications/OptimizationApplication/custom_utilities/controls/simp_utils.cpp:23]: What does simp stand for?

File: applications/OptimizationApplication/custom_utilities/controls/simp_utils.cpp [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+                return i;
+            }
+            else if (Y == rYLimits[i + 1]) {
+                return i + 1;
+            }
+    }
+    return 0;
+}
+
+double SimpUtils::ProjectionDerivative(
```
Comment (FMeister) [inline applications/OptimizationApplication/custom_utilities/controls/simp_utils.cpp:71]: Same here. Shouldn't it read "ProjectDerivativeOnto....."

File: applications/OptimizationApplication/python_scripts/controls/simp_control.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
+import KratosMultiphysics as Kratos
+import KratosMultiphysics.OptimizationApplication as KratosOA
+from KratosMultiphysics.OptimizationApplication.controls.control import Control
+from KratosMultiphysics.OptimizationApplication.optimization_info import OptimizationInfo
+from KratosMultiphysics.OptimizationApplication.utilities.response_function_implementor import ResponseFunctionImplementor
+from KratosMultiphysics.OptimizationApplication.utilities.helper_utilities import WriteCollectiveVariableDataHolderToOptmizationInfo
```
Comment (FMeister) [inline applications/OptimizationApplication/python_scripts/controls/simp_control.py:6]:
```suggestion
from KratosMultiphysics.OptimizationApplication.utilities.helper_utilities import WriteCollectiveVariableDataHolderToOptimizationInfo
```
Typo

File: applications/OptimizationApplication/python_scripts/controls/simp_control.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+        # creates element specific properties
+        if "model_parts_with_element_specific_properties" not in self.optimization_info.GetSolutionStepData(0).keys():
+            self.optimization_info["model_parts_with_element_specific_properties"] = []
+
+        for model_part in self.model_parts:
+            if not f"{model_part.FullName()}.Elements" in self.optimization_info["model_parts_with_element_specific_properties"]:
+                KratosOA.OptimizationUtils.CreateEntitySpecificPropertiesForContainer(model_part, model_part.Elements)
+                self.optimization_info["model_parts_with_element_specific_properties"].append(f"{model_part.FullName()}.Elements")
+
+                Kratos.Logger.PrintInfo(self.__class__.__name__, f"Created entity pecific properties for {model_part.FullName()}.")
```
Comment (FMeister) [inline applications/OptimizationApplication/python_scripts/controls/simp_control.py:51]:
```suggestion
                Kratos.Logger.PrintInfo(self.__class__.__name__, f"Created entity specific properties for {model_part.FullName()}.")
```
Typo

File: applications/OptimizationApplication/python_scripts/controls/simp_control.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+        for phi in self.phi_collective_container_data.GetVariableDataHolders():
+            key_refix =  f"problem_data/control_data/{phi.GetModelPart().FullName()}/{self.GetName()}"
+
+            # write density for post processing
+            density = KratosOA.ElementPropertiesContainerVariableDataHolder(phi.GetModelPart())
+            density.ReadDataFromContainerVariable(Kratos.DENSITY)
+            self.optimization_info.SetValue(f"{key_refix}/DENSITY", density)
+
+            # write young modulus for post processing
+            yound_modulus = KratosOA.ElementPropertiesContainerVariableDataHolder(phi.GetModelPart())
```
Comment (FMeister) [inline applications/OptimizationApplication/python_scripts/controls/simp_control.py:85]:
```suggestion
            young_modulus = KratosOA.ElementPropertiesContainerVariableDataHolder(phi.GetModelPart())
```
Typo

**Review by FMeister — COMMENTED** [review]
Only small problems.

File: applications/OptimizationApplication/python_scripts/controls/simp_control.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
+import KratosMultiphysics as Kratos
+import KratosMultiphysics.OptimizationApplication as KratosOA
+from KratosMultiphysics.OptimizationApplication.controls.control import Control
+from KratosMultiphysics.OptimizationApplication.optimization_info import OptimizationInfo
+from KratosMultiphysics.OptimizationApplication.utilities.response_function_implementor import ResponseFunctionImplementor
+from KratosMultiphysics.OptimizationApplication.utilities.helper_utilities import WriteCollectiveVariableDataHolderToOptmizationInfo
+
+class SimpControl(Control):
```
Comment (sunethwarna) [inline applications/OptimizationApplication/python_scripts/controls/simp_control.py:8]: Move the base control class to c++, and change optimization_info to be able to use in C++

Comment (sunethwarna) [reply applications/OptimizationApplication/python_scripts/controls/simp_control.py:8]: optimization_info will be able to store `double`, `int`, `string` ,`ContainerVariableData`

**Pull request closed without merging**